dm/syncer/data_validator.go

```diff
@@ -198,3 +399,158 @@ func (v *DataValidator) Stage() pb.Stage {
 	defer v.RUnlock()
 	return v.stage
 }
+
+func (v *DataValidator) startValidateWorkers() {
+}
+
+func (v *DataValidator) dispatchRowChange(key string, row *rowChange) {
+}
+
+func (v *DataValidator) processRowsEvent(header *replication.EventHeader, ev *replication.RowsEvent) error {
+	sourceTable := &filter.Table{
+		Schema: string(ev.Table.Schema),
+		Name:   string(ev.Table.Table),
+	}
+	fullTableName := sourceTable.String()
+	if _, ok := v.unsupportedTable[fullTableName]; ok {
+		return nil
+	}
+
+	targetTable := v.syncer.route(sourceTable)
+	tableInfo, err := v.syncer.schemaTracker.GetTableInfo(sourceTable)
+	if err != nil {
+		if schema.IsTableNotExists(err) {
+			// not a table need to sync
+			return nil
+		}
+		return terror.Annotate(err, "failed to get table info")
+	}
+
+	if len(tableInfo.Columns) < int(ev.ColumnCount) {
+		v.unsupportedTable[fullTableName] = "binlog has more columns than current table"
+		return nil
+	}
+
+	var primaryIdx *model.IndexInfo
+	for _, idx := range tableInfo.Indices {
+		if idx.Primary {
+			primaryIdx = idx
+		}
+	}
+	if primaryIdx == nil {
+		// todo: for table without primary index, need to record in the failed table, will add it later together with checkpoint
+		v.unsupportedTable[fullTableName] = "without primary key"
+		return nil
+	}
+
+	table := &validateTableInfo{
+		Source:     sourceTable,
+		Info:       tableInfo,
+		PrimaryKey: primaryIdx,
+		Target:     targetTable,
+	}
+
+	for _, cols := range ev.SkippedColumns {
+		if len(cols) > 0 {
+			err := errors.Errorf("unexpected skipped columns for table %s", sourceTable.String())
+			return err
+		}
+	}
+
+	changeType := getRowChangeType(header.EventType)
+	v.changeEventCount[changeType].Inc()
+
+	columnMap := make(map[string]*model.ColumnInfo)
+	for _, col := range tableInfo.Columns {
+		columnMap[col.Name.O] = col
+	}
+	pk := table.PrimaryKey
+	pkIndices := make([]int, len(pk.Columns))
+	for i, col := range pk.Columns {
+		pkIndices[i] = columnMap[col.Name.O].Offset
+	}
+	table.pkIndices = pkIndices
+
+	step := 1
+	if changeType == rowUpdated {
+		step = 2
+	}
+	for i := 0; i < len(ev.Rows); i += step {
+		row := ev.Rows[i]
+		pkValue := make([]string, len(pk.Columns))
+		for _, idx := range pkIndices {
+			pkValue[idx] = string(genColData(row[idx]))
+		}
+		key := genRowKey(pkValue)
+
+		if changeType == rowUpdated {
+			afterRowChangeType := changeType
+			afterRow := ev.Rows[i+1]
+			afterPkValue := make([]string, len(pk.Columns))
+			for _, idx := range pkIndices {
+				afterPkValue[idx] = string(genColData(afterRow[idx]))
+			}
+			afterKey := genRowKey(afterPkValue)
+			if afterKey != key {
+				// convert to delete and insert
```

Comment:
> // convert to delete and insert

this is similar to `RowChange.SplitUpdate()`, and checking if PK is changed is similar to `RowChange.IsIdentityUpdated()`. I really hope we can use one representation and derive many methods as needed from it. Please left a TODO for it